File: 2024/day2.py
from utils import get_file_output

output = get_file_output("day2.txt")
lines = []
for line in output:
    lines.append(line.strip().split(" "))

def is_line_safe(arr):
    unsafe = False
    first_num = int(arr[0])
    second_num = int(arr[1])
    increasing = first_num < second_num

    i = 0
    while i < len(arr):
        if i == len(arr) - 1:
            break

        this_num = int(arr[i])
        next_num = int(arr[i+1])
        if increasing:
            if this_num < next_num:
                if 1 <= next_num - this_num <= 3:
                    ...
                else:
                    unsafe = True
                    break
            else:
                unsafe = True
                break
        else:
            if this_num > next_num:
                if 1 <= this_num - next_num <= 3:
                    ...
                else:
                    unsafe = True
                    break
            else:
                unsafe = True
                break

        i += 1

    return not unsafe


def part1():
    num_safe = 0
    num_unsafe = 0
    for arr in lines:
        safe = is_line_safe(arr)

        if safe:
            num_safe += 1
        else:
            num_unsafe += 1

    print("safe: ", num_safe)
    print("unsafe: ", num_unsafe)


def unsafe_retry(starting_idx, arr):
    while starting_idx < len(arr):
        new_arr = arr.copy()
        new_arr.pop(starting_idx)

        safe = is_line_safe(new_arr)
        if safe:
            return True
        else:
            starting_idx += 1

    return False


def part2():
    num_safe = 0
    num_unsafe = 0
    for arr in lines:
        safe = is_line_safe(arr)

        if not safe:
            safe = unsafe_retry(0, arr)

        if safe:
            num_safe += 1
        else:
            num_unsafe += 1

    print("safe: ", num_safe)
    print("unsafe: ", num_unsafe)


if __name__ == '__main__':
    part1()
    part2()

File: 2024/golang/day2/day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func getOutput() [][]string {
	f, err := os.Open("../../input/day2.txt")

	check(err)
	scanner := bufio.NewScanner(f)
	list1 := make([][]string, 0)
	// list2 := make([]int, 0)

	for scanner.Scan() {
		line := scanner.Text()
		trimmed_line := strings.Trim(line, " ")
		split_line := strings.Split(trimmed_line, " ")

		list1 = append(list1, split_line)
	}

	return list1
}

func makeInts(arr []string) []int {
	new_arr := make([]int, len(arr))
	for idx, num := range arr {
		casted_number, _ := strconv.Atoi(num)
		new_arr[idx] = casted_number
	}

	return new_arr
}

func partOne() {
	num_safe := 0
	num_unsafe := 0

	for _, line := range getOutput() {
		casted_output := makeInts(line)

		safe := isLineSafe(casted_output)
		if safe {
			num_safe += 1
		} else {
			num_unsafe += 1
		}
	}

	fmt.Printf("safe: %d\n", num_safe)
	fmt.Printf("unsafe: %d\n", num_unsafe)
}

func partTwo() {
	num_safe := 0
	num_unsafe := 0

	for _, line := range getOutput() {
		casted_output := makeInts(line)

		safe := isLineSafe(casted_output)

		if !safe {
			safe = unsafeRetry(0, casted_output)
		}
		if safe {
			num_safe += 1
		} else {
			num_unsafe += 1
		}
	}

	fmt.Printf("safe: %d\n", num_safe)
	fmt.Printf("unsafe: %d\n", num_unsafe)
}

func main() {
	partOne()
	partTwo()
}
